### PR TITLE
tools: improved newgtlds.go, removed replace-between.

### PR DESCRIPTION
**Pull request by cpu**
#### tools: improved newgtlds.go, removed replace-between. - https://github.com/publicsuffix/list/commit/d996dadc0c00bdadbbebfa26e3a38fd1c3dc5a7e

Previously the `tools/newgtlds.go` utility was difficult to use with the new Github action based pull-request workflow because it _always_ updated the timestamp in the header comment in the gTLD section of `public_suffix_list.dat` when it was run, even if no gTLD data changed. This meant there was a diff in the workdir after the tool ran that would cause a PR to be opened by the action. See https://github.com/publicsuffix/list/pull/1166 for more discussion/background.

In the old travis version of the automation I used a crude shell pipeline to exclude the header comment when deciding if there was a diff or not. With the new action it's not possible to change how the diff status is determined (without re-inventing the pull request action in-repo). So instead we need to abandon the timestamp in the gTLD section header comment (that would be sad) or, we need the script to be idempotent when there's no true data change.

To make the tooling smarter about when it makes changes I reworked `newgtlds.go` to be able to consume the existing `public_suffix_list.dat` and to split apart the data templating and the header comment templating. Now the tool can exclude the header comment and compare against the existing data to determine if there is a real data change or not. If there's no change in the data then the file is left untouched. If there is a change then the file is updated and only then is the header comment with the date stamp updated along with the new data. I've included unit tests that get 80% statement coverage of this new tooling and also tested it locally with success. Conveniently this rework also lets us remove the `tools/replace-between` Perl script reducing the number of languages in play for this process.

#### CI: re-enable github tld-update workflow - https://github.com/publicsuffix/list/commit/785683990adb2da19f335c5e33fa7a411be8cbc6

Now that the tooling won't produce PRs without meaningful data updates we can re-enable the `tld-update.yml` workflow.

#### CI: run go unit tests in tld-update workflow - https://github.com/publicsuffix/list/commit/f6749a5e9c729f9f36f126ee8920c57083c5921f

Before using the `tools/newgtlds.go` tooling to open a PR we should run the unit tests to make sure the code works the way we expect. This will also help catch any bitrot.

### CODE REVIEW

**Comment by cpu** [comment]
FWIW, running `tools/patchnewgtlds` with this branch produces the same diff to `public_suffix_go.dat` as https://github.com/publicsuffix/list/pull/1203, which I opened by manually kicking the Travis automation while working on this code.

**Comment by cpu** [comment]
@sleevi @dnsguru If you folks could give this a :mag: when time permits I would appreciate it. The Travis automation is only running when I remember to kick it manually so I'm eager to get the Github workflows working again :-)

**Comment by cpu** [comment]
One last note: it's _probably_ time to split the one `.go` file (and it's test `_test.go` file) into a package with a few smaller `.go` files. I avoided doing that in this PR to make it easier to see what code was unchanged. Happy to do some additional cleanup in a follow-up.

File: tools/newgtlds.go [hunk 1]
```diff
@@ -131,6 +148,161 @@ func (e pslEntry) Comment() string {
...
+	pslDatLines := strings.Split(pslData, "\n")
+
+	headerIndex, footerIndex := 0, 0
+	for i := 0; i < len(pslDatLines); i++ {
+		if line := pslDatLines[i]; line == PSL_GTLDS_SECTION_HEADER && headerIndex == 0 {
+			headerIndex = i
+		} else if line == PSL_GTLDS_SECTION_FOOTER && footerIndex == 0 {
+			footerIndex = i
+		}
+		if headerIndex != 0 && footerIndex != 0 {
```
[the remark this answers is not in the document]
Comment (cpu) [reply tools/newgtlds.go:281]: > That is, if you have header, header, footer, treat it as an error?

That sounds sensible to me. :+1:

**Comment by cpu** [comment]
> I'm not the best for Golang reviews, but if @weppos and @dnsguru aren't comfortable, more than happy to take a second pass :)

@sleevi If it would be helpful I can probably beg another Go-savvy 3rd party to take a peek too. LMK.

> Structurally it LG (and yes, agreed on splitting). Although I hate Perl and love to see it go, I'm a little sad by how much Go-code is needed to replace the Regex :P

Haha. It's hard to win a fight with Perl on succinctness, especially in the domain of string manip and with a language as verbose as Go :muscle: :camel: :muscle: Hopefully unit tests + a little bit more care with error cases justifies the extra code to review.

File: tools/newgtlds.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (cpu) [reply tools/newgtlds.go:296]: Implemented (w/ test) in 3e05fdf

**Comment by cpu** [comment]
> FWIW, running tools/patchnewgtlds with this branch produces the same diff to public_suffix_go.dat as #1203, which I opened by manually kicking the (soon-to-be-replaced) Travis automation while working on this code.

And, most importantly, after pulling in the updated `.dat` in  e866da0 running the tooling again produces **no diff** because there haven't been any subsequent updates to the upstream JSON:

```
list on  cpu-fix-nop-prs [$?]
> git diff && echo $status
0

list on  cpu-fix-nop-prs [$?]
> git rev-parse HEAD
3e05fdf736be215e237cca951b259857feb32ddb

list on  cpu-fix-nop-prs [$?]
> bash tools/patchnewgtlds && git diff && echo $status
0
```

**Comment by weppos** [comment]
I have not investigated yet why, but tests are not passing locally to me

```
➜  publicsuffix-list git:(cpu-fix-nop-prs) go test tools/*.go
got content:

...

// newGTLDs

// List of new gTLDs imported from https://www.icann.org/resources/registries/gtlds/v2/gtlds.json on 2021-02-07T15:58:00Z
// This list is auto-generated, don't edit it manually.
// aaa : 2015-02-26 American Automobile Association, Inc.
aaa

// accountants : 2014-03-20 Binky Moon, LLC
accountants


// ===END ICANN DOMAINS===

...
expected content:

...

// newGTLDs

// List of new gTLDs imported from https://www.icann.org/resources/registries/gtlds/v2/gtlds.json on 2021-02-07T15:58:00-05:00
// This list is auto-generated, don't edit it manually.
// aaa : 2015-02-26 American Automobile Association, Inc.
aaa

// accountants : 2014-03-20 Binky Moon, LLC
accountants


// ===END ICANN DOMAINS===

...
--- FAIL: TestProcess (0.00s)
    --- FAIL: TestProcess/change_in_data (0.00s)
        newgtlds_test.go:806: expected content: "\n...\n\n// newGTLDs\n\n// List of new gTLDs imported from https://www.icann.org/resources/registries/gtlds/v2/gtlds.json on 2021-02-07T15:58:00-05:00\n// This list is auto-generated, don't edit it manually.\n// aaa : 2015-02-26 American Automobile Association, Inc.\naaa\n\n// accountants : 2014-03-20 Binky Moon, LLC\naccountants\n\n\n// ===END ICANN DOMAINS===\n\n...\n", got "\n...\n\n// newGTLDs\n\n// List of new gTLDs imported from https://www.icann.org/resources/registries/gtlds/v2/gtlds.json on 2021-02-07T15:58:00Z\n// This list is auto-generated, don't edit it manually.\n// aaa : 2015-02-26 American Automobile Association, Inc.\naaa\n\n// accountants : 2014-03-20 Binky Moon, LLC\naccountants\n\n\n// ===END ICANN DOMAINS===\n\n...\n"
FAIL
FAIL	command-line-arguments	0.547s
FAIL
```

**Comment by weppos** [comment]
It looks like a timestamp issue.

Expected:

```
on 2021-02-07T15:58:00-05:00\n//
```

Got

```
on 2021-02-07T15:58:00Z\n//
```

**Comment by weppos** [comment]
https://github.com/publicsuffix/list/blob/6424a81cc07a87b1305a6197ad5ec50e0c808ac4/tools/newgtlds_test.go#L679-L681

I admit I am a bit confused and confused as to why it passes the CI with a UTC-5 timezone.

**Comment by cpu** [comment]
@weppos Thanks for taking a look (and fixing my typos 😆 ). I'll push some revisions later tonight.

> It looks like a timestamp issue.

Oh! Yes, I should switch everything to UTC. Good catch.

> I admit I am a bit puzzled as to how it passes the CI with a UTC-5 timezone.

I think I failed to notice the current Travis config doesn't run the Go tests :-( I'll take a look at adding this in.

**Comment by weppos** [comment]
> I think I failed to notice the current Travis config doesn't run the Go tests :-( I'll take a look at adding this in.

Oh, that explains why it was all green then :D

**Comment by cpu** [comment]
> I should switch everything to UTC. Good catch.

@weppos I think with the latest two commits I pushed the tests should pass for you locally.

> I think I failed to notice the current Travis config doesn't run the Go tests :-( I'll take a look at adding this in.

Still working on this part, presently stymied by the long Travis build queues :weary: Should have another commit to push shortly.

**Comment by cpu** [comment]
>> I think I failed to notice the current Travis config doesn't run the Go tests :-( I'll take a look at adding this in.
> Still working on this part, presently stymied by the long Travis build queues weary Should have another commit to push shortly.

Ok, got the tests running in CI with 11b7e96.

> @weppos I think with the latest two commits I pushed the tests should pass for you locally.

I may have spoken too soon. My first CI build failed with a mismatched date again: https://travis-ci.org/github/cpu/list/builds/758293996 I will spend some more time investigating. :facepalm:

**Comment by cpu** [comment]
> I will spend some more time investigating. :facepalm:

All set as of c135b1b :green_circle: :ballot_box_with_check:

**Review by weppos — APPROVED** [review]
Thanks, I confirm it works as expected 👏 

I also agree with the changes to add the tests under CI.

**Comment by weppos** [comment]
Thanks for all the work @cpu . Regarding @sleevi note about how much Go is required, that is true and part of me blames the current unstructured format of the list which makes it even more challenging.

I confess I have been considering more than once to propose to switch to a structured format (e.g. YAML) and post-compile the TXT-based `.dat` version for backward compatibility.

**Comment by cpu** [comment]
> I confess I have been considering more than once to propose to switch to a structured format (e.g. YAML) and post-compile the TXT-based .dat version for backward compatibility.

That does sound appealing :+1:

**Comment by cpu** [comment]
Do you folks want me to squash my review commits and force push, or will you squash merge?

**Comment by sleevi** [comment]
I’m team squash merge. Thanks for the reminder :)

**Comment by cpu** [comment]
:wave: Is this PR waiting on any additional reviewers, or just someone with the time to squash merge?

**Comment by weppos** [comment]
> 👋 Is this PR waiting on any additional reviewers, or just someone with the time to squash merge?

Done :)

**Comment by cpu** [comment]
Thanks! I'll keep an eye on the action over the next while to make sure it's working as expected.

**Comment by cpu** [comment]
> Edit: the first run seems to have worked correctly and not created a PR with no meaningful data differences. Once a real update occurs we should see the full thing working.

We have our first data update PR: https://github.com/publicsuffix/list/pull/1214

It looks like it was successful in the sense that the PR was made, but I did see one small error in the [action output](https://github.com/publicsuffix/list/runs/1938957607?check_suite_focus=true):

> Applying labels 'tld-update'
> Error: Must have admin rights to Repository.

This is a hold-over from the ZLInt version. I'll make a follow-up PR to remove the tagging feature since it definitely doesn't justify admin rights to the repo :grimacing: 
https://github.com/publicsuffix/list/blob/fad56fbe411f7a56539527c62f54b0e124bb7d0e/.github/workflows/tld-update.yml#L43

**Comment by cpu** [comment]
>  I'll make a follow-up PR to remove the tagging feature since it definitely doesn't justify admin rights to the repo grimacing

https://github.com/publicsuffix/list/pull/1215

**Pull request merged**